libs/langchain-community/src/tools/puppeteer.ts

```diff
@@ -0,0 +1,181 @@
+  await page.goto(baseUrl, { waitUntil: "networkidle0" });
+
+  const bodyHtml = await page.content();
+  await browser.close();
+
+  return bodyHtml;
+};
+
+const DEFAULT_HEADERS = {
+  Accept:
+    "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8",
+  "Accept-Encoding": "gzip, deflate",
+  "Accept-Language": "en-US,en;q=0.5",
+
+  Referer: "https://www.google.com/",
+  "Sec-Fetch-Dest": "document",
+  "Sec-Fetch-Mode": "navigate",
+  "Sec-Fetch-Site": "cross-site",
+  "Upgrade-Insecure-Requests": "1",
+  "User-Agent":
+    "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/111.0",
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Headers = Record<string, any>;
+
+/**
+ * Defines the arguments that can be passed to the WebBrowser constructor.
+ * It extends the ToolParams interface and includes properties for a
+ * language model, embeddings, HTTP headers, and a text splitter.
+ */
+export interface PuppeteerBrowserArgs extends ToolParams {
+  model: LanguageModelLike;
+
+  embeddings: EmbeddingsInterface;
+
+  headers?: Headers;
+
+  textSplitter?: TextSplitter;
+}
+
+export class PuppeteerBrowser extends Tool {
+  name = "puppeteer-browser";
+
+  description = `Useful for when you need to find something on a webpage. Input should be a comma separated list of "ONE valid URL including protocol","What you want to find on the page or empty string for a summary"`;
+
+  static lc_name() {
+    return "PuppeteerBrowser";
+  }
+
+  get lc_namespace() {
+    return [...super.lc_namespace, "puppeteer"];
+  }
+
+  private model: LanguageModelLike;
+
+  private embeddings: EmbeddingsInterface;
+
+  private headers: Headers;
+
+  private textSplitter: TextSplitter;
+
+  constructor({
+    model,
+    headers,
+    embeddings,
+    textSplitter,
+  }: PuppeteerBrowserArgs) {
+    super(...arguments);
+
+    this.model = model;
+    this.embeddings = embeddings;
+    this.headers = headers ?? DEFAULT_HEADERS;
+
+    this.textSplitter =
+      textSplitter ??
+      new RecursiveCharacterTextSplitter({
+        chunkSize: 2000,
+        chunkOverlap: 200,
+      });
+  }
+
+  async _call(inputs: string) {
+    const [baseUrl, task] = parseInputs(inputs);
+    const doSummary = !task;
+
+    let text;
+    try {
+      const html = await getHtml(baseUrl, this.headers);
+      text = await getRelevantHtml(html);
+    } catch (e) {
+      if (e) {
+        return e.toString();
+      }
+      return "There was a problem connecting to the site";
+    }
+
+    const texts = await this.textSplitter.splitText(text);
+
+    let context;
+    if (doSummary) {
+      context = texts.slice(0, 4).join("\n");
+    } else {
+      const docs = texts.map(
+        (pageContent: string) =>
+          new Document({
+            pageContent,
+            metadata: [],
+          })
+      );
+
+      const vectorStore = await MemoryVectorStore.fromDocuments(
```

Comment:
Consider using contextual compression - likely no need for an explicit vectorstore:

https://js.langchain.com/v0.2/docs/how_to/contextual_compression

https://github.com/langchain-ai/weblangchain/blob/main/nextjs/app/api/chat/stream_log/route.ts#L101